File: core/metrics.py
from __future__ import annotations

import time
from typing import Dict, List, Optional
from flowforge.core.tracing import get_logger

logger = get_logger("metrics")

_prometheus_available = False
try:
    from prometheus_client import Counter, Histogram, Gauge, generate_latest, CollectorRegistry
    _prometheus_available = True
except ImportError:
    pass

if _prometheus_available:
    _registry = CollectorRegistry()

    tasks_total = Counter(
        "flowforge_tasks_total", "Total tasks created",
        ["mode", "status"], registry=_registry,
    )
    execution_duration = Histogram(
        "flowforge_execution_duration_seconds", "Task execution duration",
        ["mode"], buckets=[0.5, 1, 2, 5, 10, 30, 60, 120, 300],
        registry=_registry,
    )
    token_usage = Counter(
        "flowforge_token_usage_total", "Token usage",
        ["model", "provider"], registry=_registry,
    )
    tool_calls = Counter(
        "flowforge_tool_calls_total", "Tool call count",
        ["tool_name", "status"], registry=_registry,
    )
    tool_duration = Histogram(
        "flowforge_tool_duration_seconds", "Tool call duration",
        ["tool_name"], buckets=[0.1, 0.5, 1, 2, 5, 10],
        registry=_registry,
    )
    llm_errors = Counter(
        "flowforge_llm_errors_total", "LLM error count",
        ["provider", "error_type"], registry=_registry,
    )
    persona_running = Gauge(
        "flowforge_persona_running", "Currently running tasks per persona",
        ["persona"], registry=_registry,
    )

    def record_tool_call(tool_name: str, duration: float):
        tool_calls.labels(tool_name=tool_name, status="success").inc()
        tool_duration.labels(tool_name=tool_name).observe(duration)
        if tool_name not in _tool_call_data:
            _tool_call_data[tool_name] = []
        _tool_call_data[tool_name].append(duration)
        logger.debug(f"tool_call recorded: {tool_name} duration={duration:.3f}s")

    def record_tool_error(tool_name: str):
        tool_calls.labels(tool_name=tool_name, status="error").inc()
        if tool_name not in _tool_call_data:
            _tool_call_data[tool_name] = []
        _tool_call_data[tool_name].append(-1.0)

    def record_llm_tokens(provider: str, model: str, tokens: int):
        token_usage.labels(model=model, provider=provider).inc(tokens)
        key = f"{provider}/{model}"
        _llm_token_data[key] = _llm_token_data.get(key, 0) + tokens
        logger.debug(f"llm_tokens recorded: {provider}/{model} tokens={tokens}")

    def record_llm_error(provider: str, error_type: str):
        llm_errors.labels(provider=provider, error_type=error_type).inc()
        logger.debug(f"llm_error recorded: {provider} error_type={error_type}")

    def record_task_created(mode: str, persona: str):
        tasks_total.labels(mode=mode, status="created").inc()
        key = f"{mode}/{persona}"
        _task_created_data[key] = _task_created_data.get(key, 0) + 1
        logger.debug(f"task_created recorded: {mode}/{persona}")

    def record_task_completed(mode: str, persona: str, duration: float):
        tasks_total.labels(mode=mode, status="completed").inc()
        execution_duration.labels(mode=mode).observe(duration)
        key = f"{mode}/{persona}"
        _task_completed_data[key] = _task_completed_data.get(key, 0) + 1
        if key not in _task_durations_data:
            _task_durations_data[key] = []
        _task_durations_data[key].append(duration)
        logger.debug(f"task_completed recorded: {mode}/{persona} duration={duration:.3f}s")

    def record_task_failed(mode: str, persona: str):
        tasks_total.labels(mode=mode, status="failed").inc()
        key = f"{mode}/{persona}"
        _task_failed_data[key] = _task_failed_data.get(key, 0) + 1
        logger.debug(f"task_failed recorded: {mode}/{persona}")

    def set_persona_running(persona: str, count: int):
        persona_running.labels(persona=persona).set(count)

    def get_prometheus_metrics() -> bytes:
        return generate_latest(_registry)

    def get_metrics() -> dict:
        return {
            "tool_stats": get_tool_stats(),
            "task_stats": get_task_stats(),
            "llm_token_stats": get_llm_token_stats(),
        }

    _tool_call_data: Dict[str, List[float]] = {}
    _task_created_data: Dict[str, int] = {}
    _task_completed_data: Dict[str, int] = {}
    _task_failed_data: Dict[str, int] = {}
    _task_durations_data: Dict[str, List[float]] = {}
    _llm_token_data: Dict[str, int] = {}

    def get_tool_stats() -> Dict[str, dict]:
        stats = {}
        for name, durations in _tool_call_data.items():
            if not durations:
                continue
            success_durations = [d for d in durations if d >= 0]
            error_count = sum(1 for d in durations if d < 0)
            entry: dict = {
                "call_count": len(durations),
                "error_count": error_count,
            }
            if success_durations:
                entry.update({
                    "total_duration": sum(success_durations),
                    "avg_duration": sum(success_durations) / len(success_durations),
                    "min_duration": min(success_durations),
                    "max_duration": max(success_durations),
                })
            stats[name] = entry
        return stats

    def get_task_stats() -> Dict[str, dict]:
        stats = {}
        all_keys = set(_task_created_data) | set(_task_completed_data) | set(_task_failed_data)
        for key in all_keys:
            durations = _task_durations_data.get(key, [])
            stats[key] = {
                "created": _task_created_data.get(key, 0),
                "completed": _task_completed_data.get(key, 0),
                "failed": _task_failed_data.get(key, 0),
                "avg_duration": sum(durations) / len(durations) if durations else 0,
            }
        return stats

    def get_llm_token_stats() -> Dict[str, int]:
        return dict(_llm_token_data)

else:
    _tool_call_durations: Dict[str, List[float]] = {}
    _tool_error_counts: Dict[str, int] = {}
    _llm_token_counts: Dict[str, int] = {}
    _llm_error_counts: Dict[str, Dict[str, int]] = {}

    def record_tool_call(tool_name: str, duration: float):
        if tool_name not in _tool_call_durations:
            _tool_call_durations[tool_name] = []
        _tool_call_durations[tool_name].append(duration)
        logger.debug(f"tool_call recorded: {tool_name} duration={duration:.3f}s")

    def record_tool_error(tool_name: str):
        _tool_error_counts[tool_name] = _tool_error_counts.get(tool_name, 0) + 1

    def record_llm_tokens(provider: str, model: str, tokens: int):
        key = f"{provider}/{model}"
        _llm_token_counts[key] = _llm_token_counts.get(key, 0) + tokens
        logger.debug(f"llm_tokens recorded: {key} tokens={tokens}")

    def record_llm_error(provider: str, error_type: str):
        if provider not in _llm_error_counts:
            _llm_error_counts[provider] = {}
        _llm_error_counts[provider][error_type] = _llm_error_counts[provider].get(error_type, 0) + 1
        logger.debug(f"llm_error recorded: {provider} error_type={error_type}")

    _task_created: Dict[str, int] = {}
    _task_completed: Dict[str, int] = {}
    _task_failed: Dict[str, int] = {}
    _task_durations: Dict[str, List[float]] = {}
    _persona_running_counts: Dict[str, int] = {}

    def record_task_created(mode: str, persona: str):
        key = f"{mode}/{persona}"
        _task_created[key] = _task_created.get(key, 0) + 1
        logger.debug(f"task_created recorded: {key}")

    def record_task_completed(mode: str, persona: str, duration: float):
        key = f"{mode}/{persona}"
        _task_completed[key] = _task_completed.get(key, 0) + 1
        if key not in _task_durations:
            _task_durations[key] = []
        _task_durations[key].append(duration)
        logger.debug(f"task_completed recorded: {key} duration={duration:.3f}s")

    def record_task_failed(mode: str, persona: str):
        key = f"{mode}/{persona}"
        _task_failed[key] = _task_failed.get(key, 0) + 1
        logger.debug(f"task_failed recorded: {key}")

    def set_persona_running(persona: str, count: int):
        _persona_running_counts[persona] = count

    def get_prometheus_metrics() -> bytes:
        return b""

    def get_tool_stats() -> Dict[str, dict]:
        stats = {}
        for name, durations in _tool_call_durations.items():
            if not durations:
                continue
            stats[name] = {
                "call_count": len(durations),
                "total_duration": sum(durations),
                "avg_duration": sum(durations) / len(durations),
                "min_duration": min(durations),
                "max_duration": max(durations),
                "error_count": _tool_error_counts.get(name, 0),
            }
        for name in _tool_error_counts:
            if name not in stats:
                stats[name] = {
                    "call_count": 0,
                    "error_count": _tool_error_counts[name],
                }
        return stats

    def get_task_stats() -> Dict[str, dict]:
        stats = {}
        all_keys = set(_task_created) | set(_task_completed) | set(_task_failed)
        for key in all_keys:
            durations = _task_durations.get(key, [])
            stats[key] = {
                "created": _task_created.get(key, 0),
                "completed": _task_completed.get(key, 0),
                "failed": _task_failed.get(key, 0),
                "avg_duration": sum(durations) / len(durations) if durations else 0,
            }
        return stats

    def get_llm_token_stats() -> Dict[str, int]:
        return dict(_llm_token_counts)

    def get_metrics() -> dict:
        return {
            "tool_stats": get_tool_stats(),
            "task_stats": get_task_stats(),
            "llm_token_stats": get_llm_token_stats(),
        }


# ---------------------------------------------------------------------------
# MetricsCollector — 单任务指标采集器
# ---------------------------------------------------------------------------

_metrics_collectors: Dict[str, MetricsCollector] = {}


class MetricsCollector:
    """采集单个任务执行过程中的各项指标。"""

    def __init__(self, task_id: str) -> None:
        self.task_id: str = task_id
        self.start_time: float = time.time()
        self.end_time: float = 0.0
        self.llm_calls: int = 0
        self.tool_calls: int = 0
        self.tokens_in: int = 0
        self.tokens_out: int = 0
        self.cost: float = 0.0
        self.steps_total: int = 0
        self.steps_completed: int = 0
        self.errors: List[str] = []

    # -- 记录方法 ----------------------------------------------------------

    def record_llm_call(self, tokens_in: int, tokens_out: int, cost: float) -> None:
        """记录一次 LLM 调用。"""
        self.llm_calls += 1
        self.tokens_in += tokens_in
        self.tokens_out += tokens_out
        self.cost += cost
        logger.debug(
            f"MetricsCollector[{self.task_id}] llm_call: "
            f"tokens_in={tokens_in}, tokens_out={tokens_out}, cost={cost:.6f}"
        )

    def record_tool_call(self, tool_name: str, success: bool) -> None:
        """记录一次工具调用。"""
        self.tool_calls += 1
        status = "success" if success else "error"
        logger.debug(
            f"MetricsCollector[{self.task_id}] tool_call: "
            f"tool={tool_name}, status={status}"
        )

    def record_error(self, error_msg: str) -> None:
        """记录一条错误信息。"""
        self.errors.append(error_msg)
        logger.debug(f"MetricsCollector[{self.task_id}] error: {error_msg}")

    # -- 汇总方法 ----------------------------------------------------------

    def get_summary(self) -> Dict[str, object]:
        """返回指标汇总字典，同时将 end_time 设为当前时间。"""
        end = self.end_time if self.end_time > 0 else time.time()
        duration = end - self.start_time
        return {
            "task_id": self.task_id,
            "duration": round(duration, 3),
            "llm_calls": self.llm_calls,
            "tool_calls": self.tool_calls,
            "tokens_in": self.tokens_in,
            "tokens_out": self.tokens_out,
            "cost": round(self.cost, 6),
            "steps_total": self.steps_total,
            "steps_completed": self.steps_completed,
            "steps_failed": self.steps_total - self.steps_completed,
            "error_count": len(self.errors),
            "errors": list(self.errors),
        }

    def to_dict(self) -> Dict[str, object]:
        """将完整指标序列化为字典。"""
        return {
            "task_id": self.task_id,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "llm_calls": self.llm_calls,
            "tool_calls": self.tool_calls,
            "tokens_in": self.tokens_in,
            "tokens_out": self.tokens_out,
            "cost": self.cost,
            "steps_total": self.steps_total,
            "steps_completed": self.steps_completed,
            "errors": list(self.errors),
        }

    def finish(self) -> None:
        """标记任务结束，记录结束时间。"""
        self.end_time = time.time()


# -- 全局函数 ---------------------------------------------------------------


def get_metrics_collector(task_id: str) -> MetricsCollector:
    """获取或创建指定任务的指标采集器。"""
    if task_id not in _metrics_collectors:
        collector = MetricsCollector(task_id)
        _metrics_collectors[task_id] = collector
        logger.debug(f"MetricsCollector created for task: {task_id}")
    return _metrics_collectors[task_id]


def reset_metrics(task_id: str) -> None:
    """重置指定任务的指标采集器。"""
    if task_id in _metrics_collectors:
        del _metrics_collectors[task_id]
        logger.debug(f"MetricsCollector reset for task: {task_id}")

File: packages/forgekin/observability/src/metrics.ts
/**
 * @flowforge/forgekin-observability — T7.12 指标：轻量内存采集 + 单任务采集 + 全局统计。
 *
 * TS 重写自 `core/metrics.py` + `core/observability.py` 的 MetricsCollector：
 *   - MetricsCollector：计数器/仪表/直方图三种基础类型 + 快照 + Prometheus 文本导出
 *     （P-94：无第三方依赖的轻量导出，对应 Python 无 prometheus_client 时的降级语义）
 *   - TaskMetricsCollector：单任务指标采集器（LLM 调用/工具调用/token/成本/步数/错误）
 *   - GlobalMetrics：全局统计函数族（对齐 metrics.py 的内存降级分支 record_* / get_*）
 */

/** 指标标签（Prometheus 风格 key=value）。 */
export type MetricLabels = Record<string, string | number>;

/**
 * 轻量内存指标采集器 — 计数器 / 仪表 / 直方图。
 */
export class MetricsCollector {
  private readonly counters = new Map<string, number>();
  private readonly gauges = new Map<string, number>();
  private readonly histograms = new Map<string, number[]>();

  /** 递增计数器（带可选标签）。 */
  increment(name: string, value: number = 1.0, labels: MetricLabels = {}): void {
    const key = makeKey(name, labels);
    this.counters.set(key, (this.counters.get(key) ?? 0) + value);
  }

  /** 设置仪表值（带可选标签）。 */
  gauge(name: string, value: number, labels: MetricLabels = {}): void {
    this.gauges.set(makeKey(name, labels), value);
  }

  /** 记录直方图观测值（带可选标签）。 */
  observe(name: string, value: number, labels: MetricLabels = {}): void {
    const key = makeKey(name, labels);
    const bucket = this.histograms.get(key) ?? [];
    bucket.push(value);
    this.histograms.set(key, bucket);
  }

  /** 获取所有指标的快照（对齐 Python get_snapshot）。 */
  get_snapshot(): Record<string, unknown> {
    const histograms: Record<string, { count: number; sum: number; avg: number }> = {};
    for (const [key, values] of this.histograms) {
      histograms[key] = {
        count: values.length,
        sum: values.reduce((a, b) => a + b, 0),
        avg: values.length > 0 ? values.reduce((a, b) => a + b, 0) / values.length : 0,
      };
    }
    return {
      counters: Object.fromEntries(this.counters),
      gauges: Object.fromEntries(this.gauges),
      histograms,
    };
  }

  /**
   * 导出为 Prometheus 文本格式（P-94 轻量导出，无第三方依赖）。
   *
   * counter/gauge 按标准文本格式输出，histogram 以 `_count` / `_sum` 摘要形式输出。
   */
  export_prometheus_text(): string {
    const lines: string[] = [];
    for (const [key, value] of [...this.counters].sort()) {
      const [name, labels] = splitKey(key);
      lines.push(`# TYPE ${name} counter`);
      lines.push(`${formatMetric(name, labels)} ${value}`);
    }
    for (const [key, value] of [...this.gauges].sort()) {
      const [name, labels] = splitKey(key);
      lines.push(`# TYPE ${name} gauge`);
      lines.push(`${formatMetric(name, labels)} ${value}`);
    }
    for (const [key, values] of [...this.histograms].sort()) {
      const [name, labels] = splitKey(key);
      lines.push(`# TYPE ${name} summary`);
      lines.push(`${formatMetric(name + '_count', labels)} ${values.length}`);
      lines.push(`${formatMetric(name + '_sum', labels)} ${values.reduce((a, b) => a + b, 0)}`);
    }
    return lines.length > 0 ? lines.join('\n') + '\n' : '';
  }
}

/**
 * 单任务指标采集器 — 采集单个任务执行过程中的各项指标（对齐 metrics.py）。
 */
export class TaskMetricsCollector {
  readonly task_id: string;
  readonly start_time: number;
  end_time: number = 0;
  llm_calls: number = 0;
  tool_calls: number = 0;
  tokens_in: number = 0;
  tokens_out: number = 0;
  cost: number = 0;
  steps_total: number = 0;
  steps_completed: number = 0;
  errors: string[] = [];

  constructor(task_id: string) {
    this.task_id = task_id;
    this.start_time = Date.now();
  }

  /** 记录一次 LLM 调用（tokens 输入/输出 + 成本）。 */
  record_llm_call(tokens_in: number, tokens_out: number, cost: number): void {
    this.llm_calls += 1;
    this.tokens_in += tokens_in;
    this.tokens_out += tokens_out;
    this.cost += cost;
  }

  /** 记录一次工具调用（success 决定状态口径）。 */
  record_tool_call(_tool_name: string, success: boolean): void {
    this.tool_calls += 1;
    void success;
  }

  /** 记录一条错误信息。 */
  record_error(error_msg: string): void {
    this.errors.push(error_msg);
  }

  /** 返回指标汇总字典（end_time 未设置时按当前时间）。 */
  get_summary(): Record<string, unknown> {
    const end = this.end_time > 0 ? this.end_time : Date.now();
    return {
      task_id: this.task_id,
      duration: Math.round((end - this.start_time) * 1000) / 1000,
      llm_calls: this.llm_calls,
      tool_calls: this.tool_calls,
      tokens_in: this.tokens_in,
      tokens_out: this.tokens_out,
      cost: Math.round(this.cost * 1_000_000) / 1_000_000,
      steps_total: this.steps_total,
      steps_completed: this.steps_completed,
      steps_failed: this.steps_total - this.steps_completed,
      error_count: this.errors.length,
      errors: [...this.errors],
    };
  }

  /** 将完整指标序列化为字典。 */
  to_dict(): Record<string, unknown> {
    return {
      task_id: this.task_id,
      start_time: this.start_time,
      end_time: this.end_time,
      llm_calls: this.llm_calls,
      tool_calls: this.tool_calls,
      tokens_in: this.tokens_in,
      tokens_out: this.tokens_out,
      cost: this.cost,
      steps_total: this.steps_total,
      steps_completed: this.steps_completed,
      errors: [...this.errors],
    };
  }

  /** 标记任务结束。 */
  finish(): void {
    this.end_time = Date.now();
  }
}

/**
 * 全局统计函数族（对齐 metrics.py 无 prometheus_client 时的内存降级分支）。
 *
 * 提供 tool / llm / task / persona 四类记录的聚合统计。
 */
export class GlobalMetrics {
  private readonly toolDurations = new Map<string, number[]>();
  private readonly toolErrors = new Map<string, number>();
  private readonly llmTokens = new Map<string, number>();
  private readonly llmErrors = new Map<string, Map<string, number>>();
  private readonly tasksCreated = new Map<string, number>();
  private readonly tasksCompleted = new Map<string, number>();
  private readonly tasksFailed = new Map<string, number>();
  private readonly taskDurations = new Map<string, number[]>();
  private readonly personaRunning = new Map<string, number>();

  /** 记录一次工具调用耗时（秒）。 */
  record_tool_call(tool_name: string, duration: number): void {
    const bucket = this.toolDurations.get(tool_name) ?? [];
    bucket.push(duration);
    this.toolDurations.set(tool_name, bucket);
  }

  /** 记录一次工具调用错误。 */
  record_tool_error(tool_name: string): void {
    this.toolErrors.set(tool_name, (this.toolErrors.get(tool_name) ?? 0) + 1);
  }

  /** 记录 LLM token 用量（provider/model 聚合）。 */
  record_llm_tokens(provider: string, model: string, tokens: number): void {
    const key = `${provider}/${model}`;
    this.llmTokens.set(key, (this.llmTokens.get(key) ?? 0) + tokens);
  }

  /** 记录 LLM 错误（provider × error_type）。 */
  record_llm_error(provider: string, error_type: string): void {
    const byType = this.llmErrors.get(provider) ?? new Map<string, number>();
    byType.set(error_type, (byType.get(error_type) ?? 0) + 1);
    this.llmErrors.set(provider, byType);
  }

  /** 记录任务创建（mode/persona 聚合）。 */
  record_task_created(mode: string, persona: string): void {
    const key = `${mode}/${persona}`;
    this.tasksCreated.set(key, (this.tasksCreated.get(key) ?? 0) + 1);
  }

  /** 记录任务完成（含耗时）。 */
  record_task_completed(mode: string, persona: string, duration: number): void {
    const key = `${mode}/${persona}`;
    this.tasksCompleted.set(key, (this.tasksCompleted.get(key) ?? 0) + 1);
    const bucket = this.taskDurations.get(key) ?? [];
    bucket.push(duration);
    this.taskDurations.set(key, bucket);
  }

  /** 记录任务失败。 */
  record_task_failed(mode: string, persona: string): void {
    const key = `${mode}/${persona}`;
    this.tasksFailed.set(key, (this.tasksFailed.get(key) ?? 0) + 1);
  }

  /** 设置 persona 当前运行任务数。 */
  set_persona_running(persona: string, count: number): void {
    this.personaRunning.set(persona, count);
  }

  /** 工具统计（call_count/error_count/平均·最小·最大耗时）。 */
  get_tool_stats(): Record<string, Record<string, unknown>> {
    const stats: Record<string, Record<string, unknown>> = {};
    for (const [name, durations] of this.toolDurations) {
      stats[name] = {
        call_count: durations.length,
        total_duration: durations.reduce((a, b) => a + b, 0),
        avg_duration: durations.reduce((a, b) => a + b, 0) / durations.length,
        min_duration: Math.min(...durations),
        max_duration: Math.max(...durations),
        error_count: this.toolErrors.get(name) ?? 0,
      };
    }
    for (const [name, count] of this.toolErrors) {
      if (stats[name] === undefined) {
        stats[name] = { call_count: 0, error_count: count };
      }
    }
    return stats;
  }

  /** 任务统计（created/completed/failed + 平均耗时）。 */
  get_task_stats(): Record<string, Record<string, unknown>> {
    const stats: Record<string, Record<string, unknown>> = {};
    const keys = new Set([
      ...this.tasksCreated.keys(),
      ...this.tasksCompleted.keys(),
      ...this.tasksFailed.keys(),
    ]);
    for (const key of keys) {
      const durations = this.taskDurations.get(key) ?? [];
      stats[key] = {
        created: this.tasksCreated.get(key) ?? 0,
        completed: this.tasksCompleted.get(key) ?? 0,
        failed: this.tasksFailed.get(key) ?? 0,
        avg_duration:
          durations.length > 0
            ? durations.reduce((a, b) => a + b, 0) / durations.length
            : 0,
      };
    }
    return stats;
  }

  /** LLM token 统计（provider/model → tokens）。 */
  get_llm_token_stats(): Record<string, number> {
    return Object.fromEntries(this.llmTokens);
  }

  /** 汇总指标（tool_stats / task_stats / llm_token_stats）。 */
  get_metrics(): Record<string, unknown> {
    return {
      tool_stats: this.get_tool_stats(),
      task_stats: this.get_task_stats(),
      llm_token_stats: this.get_llm_token_stats(),
    };
  }
}

// ── 内部工具 ──────────────────────────────────────────────────────────────

/** 合成带标签的键（对齐 Python _make_key：name{k=v,k2=v2}）。 */
function makeKey(name: string, labels: MetricLabels): string {
  if (Object.keys(labels).length === 0) return name;
  const labelStr = Object.entries(labels)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}=${v}`)
    .join(',');
  return `${name}{${labelStr}}`;
}

/** 拆分复合键为 (metric_name, labels)。 */
function splitKey(key: string): [string, MetricLabels] {
  const open = key.indexOf('{');
  if (open === -1) return [key, {}];
  const name = key.slice(0, open);
  const labelsPart = key.slice(open + 1, key.endsWith('}') ? -1 : undefined);
  const labels: MetricLabels = {};
  if (labelsPart.length > 0) {
    for (const pair of labelsPart.split(',')) {
      const eq = pair.indexOf('=');
      if (eq !== -1) {
        labels[pair.slice(0, eq)] = pair.slice(eq + 1);
      }
    }
  }
  return [name, labels];
}

/** Prometheus 文本行（对齐 Python _format_metric）。 */
function formatMetric(name: string, labels: MetricLabels): string {
  if (Object.keys(labels).length === 0) return name;
  const labelStr = Object.entries(labels)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}="${v}"`)
    .join(',');
  return `${name}{${labelStr}}`;
}
